File: rowanlee92/heap-tree/CousinsInBinaryTree.java
package com.janjanee.algo.tree_heap;


import java.util.LinkedList;
import java.util.Queue;

public class CousinsInBinaryTree {

    public static boolean isCousins(TreeNode root, int x, int y) {
        Queue<TreeNode> queue = new LinkedList<>();

        queue.add(root);

        int xIndex = 0;
        int yIndex = 0;
        int index;

        while (!queue.isEmpty()) {
            int oriSize = queue.size();
            int size = queue.size();

            while (size-- > 0) {
                TreeNode node = queue.remove();
                if (node == null) continue;

                index = oriSize - size;

                if (node.val == x || node.val == y) {
                    if (xIndex == 0) {
                        xIndex = index;
                    } else {
                        yIndex = index;
                    }
                }
                queue.add(node.left);
                queue.add(node.right);

            }

            // cousins check
            if (xIndex != 0 && yIndex != 0) {
                return xIndex % 2 == 0 || yIndex != xIndex + 1;
            } else if (xIndex != 0)
                return false;
        }

        return false;
    }

}

File: yeseulo/heap-tree/cousins-in-binary-tree.js
/**
 * Leetcode -993. Cousins in Binary Tree
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} x
 * @param {number} y
 * @return {boolean}
 */
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} x
 * @param {number} y
 * @return {boolean}
 */
var isCousins = function(root, x, y) {
  const queue = [root];

  while(queue.length) {
    const size = queue.length;
    let foundX = false;
    let foundY = false;

    for (let i = 0; i < queue.length; i++) {
      const {left, right, val} = queue.shift();

      if (left && right) {
        if (
          (left.val === x && right.val === y) ||
          (left.val === y && right.val === x)
        )
        return false;
      }

      if (val === x) foundX = true;
      if (val === y) foundY = true;
      if (left) queue.push(left);
      if (right) queue.push(right);
    }

    if (foundX && foundY) {
      return true;
    }
  }

  return false;
};